plan/predicate_push_down.go

```diff
@@ -147,6 +150,107 @@ func (p *LogicalJoin) PredicatePushDown(predicates []expression.Expression) (ret
 	return ret, p.self
 }
 
+func (p *LogicalJoin) extractFiltersFromDNFs(conditions []expression.Expression) []expression.Expression {
+	var allExtracted []expression.Expression
+	for i := len(conditions) - 1; i >= 0; i-- {
+		if sf, ok := conditions[i].(*expression.ScalarFunction); ok && sf.FuncName.L == ast.LogicOr {
+			extracted, remained := p.extractFiltersFromDNF(sf)
+			allExtracted = append(allExtracted, extracted...)
+			if remained == nil {
+				conditions = append(conditions[:i], conditions[i+1:]...)
+			} else {
+				conditions[i] = remained
+			}
+		}
+	}
+	return append(conditions, allExtracted...)
+}
+
+// extractFiltersFromDNF extracts the same condition that occurs in every DNF item.
+func (p *LogicalJoin) extractFiltersFromDNF(dnfFunc *expression.ScalarFunction) ([]expression.Expression, expression.Expression) {
+	dnfItems := expression.FlattenDNFConditions(dnfFunc)
+	sc := p.ctx.GetSessionVars().StmtCtx
+	codeMap := make(map[string]int)
+	hashcode2Expr := make(map[string]expression.Expression)
+	firstRun := true
+	for _, dnfItem := range dnfItems {
+		// We need this because there may be the case like `select * from t, t1 where (t.a=t1.a and t.a=t1.a) or (something).
```

Comment:
Can we just make sure that no such thing will happen by eliminating them at an earlier stage?

Reply:
Yes, this is a better practice.